bx_py_utils/test_utils/snapshot.py

```diff
@@ -13,16 +17,73 @@
 )
 
 
+SELF_FILE_PATH = pathlib.Path(__file__)
+_AUTO_SNAPSHOT_NAME_COUNTER = Counter()
+
+
 def _write_json(obj, snapshot_file):
     with snapshot_file.open('w') as snapshot_handle:
         json.dump(obj, snapshot_handle, ensure_ascii=False, indent=4, sort_keys=True)
 
 
-def assert_text_snapshot(root_dir: Union[pathlib.Path, str], snapshot_name: str, got: str,
-                         extension: str = '.txt',
-                         fromfile: str = 'got', tofile: str = 'expected',
-                         diff_func: Callable = text_unified_diff):
-    assert re.match(r'^[-_.a-zA-Z0-9]+$', snapshot_name), f'Invalid snapshot name {snapshot_name}'
+def _get_caller_names(root_dir: Union[pathlib.Path, str] = None, snapshot_name: str = None):
+    """
+    Helper to get snapshot directory and name by stack frame info, but only if not given.
+
+    Note:
+      This only works if this function is in the same file as the snapshot functions.
+      Because we compare the file path in stack frame to get the caller frame.
+      (If this function will be moved: the file path in last_frame_by_path() must be updated, too.)
+    """
+    if root_dir:
+        if not isinstance(root_dir, pathlib.Path):
+            root_dir = pathlib.Path(root_dir)
+        assert_is_dir(root_dir)
+
+    if snapshot_name is not None:
+        assert re.match(r'^[-_.a-zA-Z0-9]+$', snapshot_name), (
+            f'Invalid snapshot name {snapshot_name!r}'
+        )
+
+    if not root_dir or not snapshot_name:
+        # Set "root_dir" and "snapshot_name" if missing by stack frame info.
+
+        # Get the caller stack frame (First frame that is outside this file path):
+        external_stack_frame = last_frame_by_path(file_path=SELF_FILE_PATH)
+        assert external_stack_frame
```

Comment:
should be moved into `last_frame_by_path`

Reply:
done